specs/phase1/beacon-chain.md

```diff
@@ -1057,51 +536,124 @@ def process_epoch(state: BeaconState) -> None:
     process_reveal_deadlines(state)
     process_challenge_deadlines(state)
     process_slashings(state)
+    process_pending_headers(state)
+    charge_confirmed_header_fees(state)
+    reset_pending_headers(state)
     process_final_updates(state)  # phase 0 final updates
     process_phase_1_final_updates(state)
 ```
 
+#### Pending headers
+
+```python
+
+def process_pending_headers(state: BeaconState):
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(SHARD_COUNT):
+            # Pending headers for this (slot, shard) combo
+            candidates = [
+                c for c in state.previous_epoch_pending_shard_headers if
+                (c.slot, c.shard) == (slot, shard)
+            ]
+            if True not in [c.confirmed for c in candidates]:
+                # The entire committee (and its balance)
+                full_committee = get_beacon_committee(state, slot, shard)
+                full_committee_balance = get_total_balance(state, full_committee)
+                # The set of voters who voted for each header
+                # (and their total balances)
+                voting_sets = [
+                    [v for i, v in enumerate(full_committee) if c.votes[i]]
+                    for c in candidates
+                ]
+                voting_balances = [
+                    get_total_balance(state, voters)
+                    for voters in voting_sets
+                ]
+                # Get the index with the most total balance voting for them.
+                # NOTE: if two choices get exactly the same voting balance,
+                # the candidate earlier in the list wins
+                if max(voting_balances) > 0:
+                    winning_index = voting_balances.index(max(voting_balances))
+                else:
+                    # If no votes, zero wins
+                    winning_index = [c.root for c in candidates].index(Root())
+                candidates[winning_index].confirmed = True
+    for slot in range(SLOTS_PER_EPOCH):
+        for shard in range(SHARD_COUNT):
+            state.most_recent_confirmed_commitments[shard][slot] = DataCommitment()
+    for c in state.previous_epoch_pending_shard_headers:
+        if c.confirmed:
+            state.most_recent_confirmed_commitments[c.shard][c.slot % SLOTS_PER_EPOCH] = c.commitment
+```            
+
+```python
+def charge_confirmed_header_fees(state: BeaconState) -> None:
+    new_gasprice = state.shard_gasprice
+    adjustment_quotient = get_active_shard_count(state) * SLOTS_PER_EPOCH * GASPRICE_ADJUSTMENT_COEFFICIENT
```

Comment:
```suggestion
    adjustment_quotient = get_active_shard_count(state, get_current_epoch(state)) * SLOTS_PER_EPOCH * GASPRICE_ADJUSTMENT_COEFFICIENT
```

Reply:
This one is a bit tricker, right? Because we would need the shard count at the time the candidate was proposed?
@vbuterin doe we have to pull the adjustment_quotient inside the loop?